File: Linked List/ReverseLinkedList2.js
/**
 * Problem: https://leetcode.com/problems/reverse-linked-list-ii/
 * Author: Dung Nguyen Tien (Chris)
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} m
 * @param {number} n
 * @return {ListNode}
 */
var reverseBetween = function(head, m, n) {
    var dummyNode = new ListNode(-1);
    dummyNode.next = head;
    
    var start = new ListNode();
    var before = dummyNode;
    var current = dummyNode.next;
    var after = new ListNode();
    var first = new ListNode;
    
    for (var i = 1; i <= n; i++) {
        // the start save the node before m, the first save the node m
        if (i >= m) {
            if (i == m) {
                start = before;
                first = current;
                console.log(start.val + "-" + first.val);
            }
            after = current.next;
            current.next = before;
            before = current;
            current = after;
        }   
        if (i == n) {
            first.next = after;
            start.next = before;
        }
        if (i < m) {
            before = current;
            current = current.next;
        }
    }
    
    return dummyNode.next;
};

File: Linked List/ReverseLinkedList2.cs
/**
 * Problem: https://leetcode.com/problems/reverse-linked-list-ii/submissions/
 * Author: Dung Nguyen Tien (Chris)
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode ReverseBetween(ListNode head, int m, int n) {
        if (head.next == null) return head;
        
        var dummyNode = new ListNode(-1);
        dummyNode.next = head;
        
        var previous = dummyNode;
        var current = dummyNode.next;
        var first = current;
        var after = current.next;
        var i = 1;
        
        while (i <= n) {
            if (i >= m) {
                if (i == m) {
                    first = current;
                }
                current.next = previous;
            }
            
            if (i == n) break;
            
            previous = current;
            current = after;
            after = after.next;
            i++;
        }
        
        var start = first.next;
        first.next = after;
        start.next = current;
        
        return dummyNode.next;
    }
}